diff --git a/cctbx_project/boost_adaptbx/meta_ext.cpp b/cctbx_project/boost_adaptbx/meta_ext.cpp
index 188c88071..a05d18d07 100644
--- a/cctbx_project/boost_adaptbx/meta_ext.cpp
+++ b/cctbx_project/boost_adaptbx/meta_ext.cpp
@@ -298,12 +298,6 @@ namespace {
 #if defined(__llvm__)
     result += "__llvm__\n";
 #endif
-#if defined(BOOST_PYTHON_HAVE_CXXABI_CXA_DEMANGLE_IS_BROKEN) \
-    && !defined(USE_CONDA) \
-    && !defined(__MINGW32__) // workaround for MinGW linking problem
-    result += "boost::python::cxxabi_cxa_demangle_is_broken(): "
-            + to_str(boost::python::cxxabi_cxa_demangle_is_broken()) + nl;
-#endif
 #if defined(__GXX_WEAK__)
     result += "__GXX_WEAK__ = " + to_str(__GXX_WEAK__) + nl;
 #endif
